### PR TITLE
`EmergencyReparentShard`: ignore non-MySQL-GTID-based replicas

### PR DESCRIPTION
**Pull request by timvaillancourt**
## Description

`EmergencyReparentShard` currently fails when a shard has a mix of GTID-based and non-GTID _(file-based or MariaDB GTID)_ replicas. This happens because `StopReplicationAndGetStatus` doesn't distinguish between replicas that use MySQL GTIDs and those that don't — non-GTID replicas can't participate in GTID-based position comparison, so they should be skipped

This PR makes ERS gracefully handle non-GTID replicas by:

1. `StopReplicationAndGetStatus` now checks if the tablet uses MySQL GTIDs _(the only supported GTID flavour)_ by inspecting the `RelayLogPosition` **before** stopping replication. Non-GTID tablets return `{Before: status, After: nil}` with a nil error — this avoids the side effect of stopping replication on tablets that will be ignored, while still making the `Before` status _(including semi-sync state)_ available to the caller through gRPC
2. `stopReplicationAndBuildStatusMaps` detects `After == nil` and either skips the tablet _(non-semi-sync)_ or fails ERS _(semi-sync acker)_. Skipped tablets are **not** added to `reachableTablets` — since replication was intentionally left running (`After == nil`), they cannot count as "reached and stopped" for the `haveRevoked` safety check
3. `restrictValidCandidates` provides a belt-and-suspenders filter for any non-GTID candidates that made it into the candidate pool
4. `hasNonZeroRelayLogPositions()` gates errant GTID detection — avoids false positives on fresh clusters with no positions to compare

The semi-sync safety check is important — if a non-GTID replica is a semi-sync acker in a GTID-based shard, it may hold the most up-to-date data that we can't position-compare with other candidates. In that case, ERS fails rather than risking data loss

Also adds `IsSemiSyncAcker()` helper on `ReplicationStatus`

## Checklist

- [x] "Backport to:" labels have been added if this change should be back-ported to release branches
- [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
- [x] Tests were added or are not required
- [x] Did the new or modified tests pass consistently locally and on CI?
- [x] Documentation was added or is not required

## Deployment Notes

No migrations or config changes. Non-GTID replicas that were previously causing ERS to fail will now be silently skipped _(unless they are semi-sync ackers, in which case ERS will fail with a clear error)_

### AI Disclosure

Claude Code assisted with development, testing, and wrote this PR summary

### CODE REVIEW

File: go/vt/vtctl/reparentutil/emergency_reparenter.go [hunk 1]
```diff
@@ -783,6 +788,18 @@ func (erp *EmergencyReparenter) filterValidCandidates(validTablets []*topodatapb
 	return notPreferredValidTablets, nil
 }
 
+// hasNonZeroRelayLogPositions checks if any candidate has non-zero relay log positions.
+// Returns false if all candidates have zero positions, which is typical for fresh clusters
+// that have never had replication set up.
+func hasNonZeroRelayLogPositions(validCandidates map[string]*RelayLogPositions) bool {
```
Comment (timvaillancourt) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go]: This check was required after moving `wrangler` tests to use MySQL GTIDs. This enabled Errant GTID checking for the first time on these test cases, which exposed a bug in downgrade CI testing when tablets have no positions to compare yet

Now we check if there are positions to compare. If there are no positions, we don't do errant detection

**Review by mattlord — COMMENTED** [review]
This feels like a lot of code to carry forward in Vitess to resolve this issue. 

This is the only callsite:
```
	// find the positions of all the valid candidates.
	validCandidates, isGTIDBased, err = FindPositionsOfAllCandidates(stoppedReplicationSnapshot.statusMap, stoppedReplicationSnapshot.primaryStatusMap)
	if err != nil {
		return err
	}
```

Can't we just ignore any non-GTID based replicas there? At least if we find a mix of GTID and non-GTID. 

Also note that we have a capability framework. I don't see this as a capability of the server, but rather a current configuration  setting / operating mode.  Remember that we do not support failovers at all w/o GTIDs and we should not attempt to. So I think we should be able to resolve the issue/bug in a much simpler and more direct way. 

Or do you think I'm missing something here?

Group Replication will be a much bigger thing so I don't think that future work should factor into our thinking here. VTOrc won't need to generally handle shard failover at all. It will need to handle some new things like e.g. loss of quorum in the group.

**Comment by mattlord** [comment]
To be more clear, I'm wondering why it may not be as simple as this? 
```
diff --git a/go/vt/vtctl/reparentutil/emergency_reparenter.go b/go/vt/vtctl/reparentutil/emergency_reparenter.go
index 3c5b0658af..65543f9503 100644
--- a/go/vt/vtctl/reparentutil/emergency_reparenter.go
+++ b/go/vt/vtctl/reparentutil/emergency_reparenter.go
@@ -159,7 +159,6 @@ func (erp *EmergencyReparenter) reparentShardLocked(ctx context.Context, ev *eve
 		validReplacementCandidates []*topodatapb.Tablet
 		betterCandidate            *topodatapb.Tablet
 		isIdeal                    bool
-		isGTIDBased                bool
 	)
 
 	shardInfo, err = erp.ts.GetShard(ctx, keyspace, shard)
@@ -216,7 +215,7 @@ func (erp *EmergencyReparenter) reparentShardLocked(ctx context.Context, ev *eve
 	}
 
 	// find the positions of all the valid candidates.
-	validCandidates, isGTIDBased, err = FindPositionsOfAllCandidates(stoppedReplicationSnapshot.statusMap, stoppedReplicationSnapshot.primaryStatusMap)
+	validCandidates, err = FindPositionsOfAllCandidates(stoppedReplicationSnapshot.statusMap, stoppedReplicationSnapshot.primaryStatusMap)
 	if err != nil {
 		return err
 	}
@@ -234,11 +233,9 @@ func (erp *EmergencyReparenter) reparentShardLocked(ctx context.Context, ev *eve
 	}
 
 	// For GTID based replication, we will run errant GTID detection.
-	if isGTIDBased {
-		validCandidates, err = erp.findErrantGTIDs(ctx, validCandidates, stoppedReplicationSnapshot.statusMap, tabletMap, opts.WaitReplicasTimeout)
-		if err != nil {
-			return err
-		}
+	validCandidates, err = erp.findErrantGTIDs(ctx, validCandidates, stoppedReplicationSnapshot.statusMap, tabletMap, opts.WaitReplicasTimeout)
+	if err != nil {
+		return err
 	}
 
 	// Find the intermediate source for replication that we want other tablets to replicate from.
diff --git a/go/vt/vtctl/reparentutil/replication.go b/go/vt/vtctl/reparentutil/replication.go
index 4777499589..2bd7b6ab0a 100644
--- a/go/vt/vtctl/reparentutil/replication.go
+++ b/go/vt/vtctl/reparentutil/replication.go
@@ -89,7 +89,7 @@ func (rlp *RelayLogPositions) IsZero() bool {
 func FindPositionsOfAllCandidates(
 	statusMap map[string]*replicationdatapb.StopReplicationStatus,
 	primaryStatusMap map[string]*replicationdatapb.PrimaryStatus,
-) (map[string]*RelayLogPositions, bool, error) {
+) (map[string]*RelayLogPositions, error) {
 	replicationStatusMap := make(map[string]*replication.ReplicationStatus, len(statusMap))
 	positionMap := make(map[string]*RelayLogPositions)
 
@@ -99,19 +99,13 @@ func FindPositionsOfAllCandidates(
 		replicationStatusMap[alias] = &status
 	}
 
-	// Determine if we're GTID-based. If we are, we'll need to look for errant
-	// GTIDs below.
-	var (
-		isGTIDBased                bool
-		isNonGTIDBased             bool
-		emptyRelayPosErrorRecorder concurrency.FirstErrorRecorder
-	)
+	var emptyRelayPosErrorRecorder concurrency.FirstErrorRecorder
 
 	for alias, status := range replicationStatusMap {
-		if _, ok := status.RelayLogPosition.GTIDSet.(replication.Mysql56GTIDSet); ok {
-			isGTIDBased = true
-		} else {
-			isNonGTIDBased = true
+		if _, ok := status.RelayLogPosition.GTIDSet.(replication.Mysql56GTIDSet); !ok {
+			// We do not support reparenting a shard to a tablet that is not GTID based.
+			delete(replicationStatusMap, alias)
+			continue
 		}
 
 		if status.RelayLogPosition.IsZero() {
@@ -122,21 +116,12 @@ func FindPositionsOfAllCandidates(
 		}
 	}
 
-	if isGTIDBased && emptyRelayPosErrorRecorder.HasErrors() {
-		return nil, false, emptyRelayPosErrorRecorder.Error()
-	}
-
-	if isGTIDBased && isNonGTIDBased {
-		return nil, false, vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "encountered mix of GTID-based and non GTID-based relay logs")
+	if emptyRelayPosErrorRecorder.HasErrors() {
+		return nil, emptyRelayPosErrorRecorder.Error()
 	}
 
 	// Store the final positions in the map.
 	for alias, status := range replicationStatusMap {
-		if !isGTIDBased {
-			positionMap[alias] = &RelayLogPositions{Combined: status.Position}
-
-			continue
-		}
 		positionMap[alias] = &RelayLogPositions{
 			Combined: status.RelayLogPosition,
 			Executed: status.Position,
@@ -146,13 +131,13 @@ func FindPositionsOfAllCandidates(
 	for alias, primaryStatus := range primaryStatusMap {
 		executedPosition, err := replication.DecodePosition(primaryStatus.Position)
 		if err != nil {
-			return nil, false, vterrors.Wrapf(err, "could not decode a primary status executed position for tablet %v: %v", alias, err)
+			return nil, vterrors.Wrapf(err, "could not decode a primary status executed position for tablet %v: %v", alias, err)
 		}
 
 		positionMap[alias] = &RelayLogPositions{Combined: executedPosition}
 	}
 
-	return positionMap, isGTIDBased, nil
+	return positionMap, nil
 }
 
 // ReplicaWasRunning returns true if a StopReplicationStatus indicates that the

```

I understand that there may be good reasons why it's not that simple, but I would like to understand. 🙂

**Comment by timvaillancourt** [comment]
@mattlord I have updated the code to remove "capability". This was added to allow non-GTID users to call the`StopReplicationAndGetStatus` RPC outside of ERS, but that flexibility isn't really needed. Now we just hard-require MySQL GTIDs 👍. That removed 7 files from the diff

> I'm wondering why it may not be as simple as this?

The only way ERS knows if a tablet is semi-sync or not is by calling `StopReplicationAndGetStatus`, which stops replication unconditionally before this PR. If we are ignoring non-MySQL-GTID-based replicas, we should only stop replication on tablets that need it to be stopped. This PR adds logic for this to happen and it's unavoidable that adds more code. I strongly feel that stopping replication on tablets we are ignoring is not a good solution

Also adding additional RPCs just to answer this is something I avoided, as a theme with reparent code is to depend on as few RPCs as possible, for reliability

This PR changes this RPC to require MySQL GTIDs and return an error that the reparent code understands if it does not. I would like to challenge that this can be done entirely in the reparent code, as without this PR, the reparent code doesn't know which tablets are semi-sync ackers or not, or if it did, it needlessly stopped replication on the non-ackers - which to me is not "ignoring"

**Review by mhamza15 — COMMENTED** [review]
I might be missing some context, but what is the harm in stopping replication on non-gtid-based replicas? Don't get me wrong, I'd love to avoid more work/room for failure as much as possible, but considering we still make the RPC, I'm not sure how much leeway that buys us.

An alternative approach might be to stop replication as usual and fail fast if any non-gtid-based replicas are active semi-sync ackers. Otherwise, we continue and exclude them from candidacy.

(I don't know if this already exists, but pairing this with a VTOrc recovery that disables semi-sync on non-gtid-based replicas might help as well.)

**Comment by timvaillancourt** [comment]
> I might be missing some context, but what is the harm in stopping replication on non-gtid-based replicas? Don't get me wrong, I'd love to avoid more work/room for failure as much as possible, but considering we still make the RPC, I'm not sure how much leeway that buys us.
> 
> An alternative approach might be to stop replication as usual and fail fast if any non-gtid-based replicas are active semi-sync ackers. Otherwise, we continue and exclude them from candidacy.
> 
> (I don't know if this already exists, but pairing this with a VTOrc recovery that disables semi-sync on non-gtid-based replicas might help as well.)

@mhamza15 the goal of this PR is to ignore non-GTID based replicas in ERS. To me that can only be achieved by not stopping replication, ie: ignoring the replica

But I'd like to flip the question around, if the goal is to ignore non-GTID-based replicas, why would we want to stop replication on non-GTID-based replicas? The intent of the stop is for future steps these replicas will never reach. Why would we want them stopped, and is this detail worth holding the PR back?

If a consensus would prefer, I can break the non-GTID-replicas replication, but to me that means the PR does not correctly achieve it's goal and is unnecessary

**Comment by timvaillancourt** [comment]
cc @mattlord on the previous comment as well

**Comment by mhamza15** [comment]
Again, lacking a bit of context, so the definition of "ignore" is a bit subjective 😅 . My understanding of the goal of this PR is to unblock ERS in shards with mixed replicas. From that lens, "ignore" to me means avoid blocking ERS if possible, and not necessarily "don't touch them at all".

To be clear, I'm not against skipping stopping replication. My thoughts are just that taking that approach seems to introduce a bit more complexity, while keeping it as is now and gating on semi-sync + filtering non-gtid-based out of `positionMap` might be simpler overall.

**Comment by timvaillancourt** [comment]
> To be clear, I'm not against skipping stopping replication. My thoughts are just that taking that approach seems to introduce a bit more complexity, while keeping it as is now and gating on semi-sync + filtering non-gtid-based out of `positionMap` might be simpler overall.

@mhamza15 the only way we can know if a tablet is non-gtid based is by calling `StopReplicationAndGetStatus`. If that RPC isn't updated to understand what we're doing, we break unrelated replicas that this PR is designed to ignore

I don't see how a few less lines is worth breaking something we are intending to ignore - we would be creating a problem just to have a few lines of less code. I don't see us making that tradeoff anywhere else in the project so I'm a bit confused, but happy to do whatever gets PRs shipped

File: go/vt/vtctl/reparentutil/emergency_reparenter.go (hunk 1 above)
Comment (mattlord) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go]: Just FYI, the wrangler code is used by the legacy deprecated vtctl/vtctlclient clients. We should remove all of it in v25 IMO.

**Review by mattlord — APPROVED** [review]
LGTM. One thing I think still feels a little off: ReparentShard() now skips errant-GTID detection when all candidate relay positions are zero, but FindPositionsOfAllCandidates() still errors _earlier_ on zero relay-log positions in GTID shards, so a true fresh-cluster / all-zero cases may still fail _before_ they reach the new guard. Do we really need the new guard? Could we either relax that earlier check for the all-zero case, or narrow the description/tests to the cases this actually handles?

I'm still a little nervous about making more than minimal changes needed to solve a problem in this area of Vitess because we've been playing whack-a-mole with edge cases here for years now. But the changes do seem correct to me.

**Review by nickvanw — COMMENTED** [review]
I found two correctness issues to address before this is safe:

1. Non-GTID replicas are added to `reachableTablets` even though replication was intentionally left running on them, which makes the revoke safety check unsound.
2. The new all-zero relay-log guard is dead code for GTID shards because `FindPositionsOfAllCandidates()` still errors before ERS reaches it.

File: go/vt/vtctl/reparentutil/replication.go [hunk 2]
```diff
@@ -304,23 +319,36 @@ func stopReplicationAndBuildStatusMaps(
...
+						"tablet %v is a non-GTID replica with semi-sync enabled; this is unsupported in a GTID-based shard", alias)
+					return
+				}
+			}
+			// The tablet was successfully contacted, so count it as reachable for the
+			// haveRevoked safety check, even though it won't be a candidate.
 			m.Lock()
-			res.tabletsBackupState[alias] = isTakingBackup
-			res.statusMap[alias] = stopReplicationStatus
 			res.reachableTablets = append(res.reachableTablets, tabletInfo.Tablet)
```
Comment (nickvanw) [inline go/vt/vtctl/reparentutil/replication.go]: This makes `haveRevoked()` treat the tablet as safe even though `After == nil` means we intentionally left replication running on it. In `DurabilityNone`, just adding the tablet to `reachableTablets` is enough for `haveRevokedForTablet()` to return true, so ERS can conclude writes are revoked while a file-based replica is still following the old primary. We need to keep these tablets out of `reachableTablets` or revoke them before counting them here.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go [hunk 3]
```diff
@@ -233,12 +235,15 @@ func (erp *EmergencyReparenter) reparentShardLocked(ctx context.Context, ev *eve
 		return err
 	}
 
-	// For GTID based replication, we will run errant GTID detection.
-	if isGTIDBased {
+	// Run errant GTID detection only if we have non-zero relay log positions.
+	// For fresh clusters with no replication history, all positions will be zero and we skip this step.
+	if hasNonZeroRelayLogPositions(validCandidates) {
```
Comment (nickvanw) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go]: I do not think this guard can fire for the fresh-cluster case described in the PR. `FindPositionsOfAllCandidates()` still records `afterStatus.RelayLogPosition.IsZero()` as an error and returns it before we get here, so a GTID shard with all-zero relay positions still fails earlier. Either relax that earlier check for the all-zero case or narrow this branch/tests to the scenarios it actually handles.

File: go/vt/vtctl/reparentutil/replication.go (hunk 2 above)
Comment (timvaillancourt) [reply go/vt/vtctl/reparentutil/replication.go]: Fixed in cdf96207ee — non-GTID tablets are no longer added to `reachableTablets`. Since replication was intentionally left running on them (`After == nil`), they can't count as "reached and stopped" for the `haveRevoked` safety check.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go (hunk 3 above)
Comment (timvaillancourt) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go]: Fixed in cdf96207ee — `FindPositionsOfAllCandidates` now only errors on zero relay-log positions when there's a *mix* of zero and non-zero positions (some tablets have data, some don't). If all positions are zero, it allows them through so the downstream `hasNonZeroRelayLogPositions()` guard correctly skips errant GTID detection.

That said, in practice this path is unlikely to fire for GTID shards since tablets always have non-zero positions after `InitShardPrimary`/`PlannedReparentShard` bootstraps the shard.

**Comment by timvaillancourt** [comment]
> LGTM. One thing I think still feels a little off: ReparentShard() now skips errant-GTID detection when all candidate relay positions are zero, but FindPositionsOfAllCandidates() still errors earlier on zero relay-log positions in GTID shards, so a true fresh-cluster / all-zero cases may still fail before they reach the new guard. Do we really need the new guard? Could we either relax that earlier check for the all-zero case, or narrow the description/tests to the cases this actually handles?

Good question! The early check in `FindPositionsOfAllCandidates()` was actually already relaxed as part of this PR to handle the all-zero case. The condition on [line 157 of replication.go](https://github.com/vitessio/vitess/blob/e737d5079c/go/vt/vtctl/reparentutil/replication.go#L157) requires `hasNonZeroRelayPos` to be true before erroring:

```go
if isGTIDBasedShard && hasNonZeroRelayPos && emptyRelayPosErrorRecorder.HasErrors() {
```

So it only errors on a *mix* of zero and non-zero positions (which indicates a real problem). If all positions are zero (fresh cluster), the early check lets it through, and the `hasNonZeroRelayLogPositions()` guard downstream then skips errant GTID detection for that case.

The two guards are complementary:
- **Mixed zero/non-zero** → early check catches it as an error
- **All zeros** → early check allows it through → later guard skips errant GTID detection
- **All non-zero** → both pass → errant GTID detection runs normally

File: go/vt/vtctl/reparentutil/replication.go [hunk 4]
```diff
@@ -338,23 +357,34 @@ func stopReplicationAndBuildStatusMaps(
...
+				if beforeStatus.IsSemiSyncAcker() {
+					err = vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION,
+						"tablet %v is a non-GTID replica with semi-sync enabled; this is unsupported in a GTID-based shard", alias)
+					return
+				}
+			}
+			// Do not add to reachableTablets: replication was intentionally left
+			// running on this tablet (After == nil), so it cannot count as
+			// "reached and stopped" for the haveRevoked safety check.
+			logger.Warningf("tablet %v does not use MySQL GTID (likely a file-based replica); skipping for emergency reparent candidate selection", alias)
```
Comment (nickvanw) [inline go/vt/vtctl/reparentutil/replication.go:396]: This treats the tablet as a clean skip, but replication is still running here. When every RPC succeeds, `stopReplicationAndBuildStatusMaps()` returns from the early `len(errRecorder.Errors) == 0` branch before it ever calls `haveRevoked()`, so ERS can proceed while a file-based replica is still following the old primary. I think this path needs to fail or explicitly revoke these tablets rather than silently skipping them.

**Review by nickvanw — COMMENTED** [review]
I still see two production correctness issues here:

1. The new `After == nil` skip path can bypass the revoke safety check on the all-success path. I left the concrete call path inline.
2. Even when a non-GTID replica is allowed through because it is not currently acking, ERS still reconfigures every replica with `policy.IsReplicaSemiSync(...)`. Under semi-sync durability that turns `REPLICA` tablets back into semi-sync ackers, and `SetReplicationSource` does not reject file-based positions, so a successful ERS can immediately recreate the unsupported mixed topology under the new primary.

I ran the targeted `go test ./go/vt/vtctl/reparentutil` cases around the new paths and they pass, but I think these two behaviors still need to be addressed before this is safe.

File: go/vt/vtctl/reparentutil/replication.go (hunk 4 above)
Comment (timvaillancourt) [reply go/vt/vtctl/reparentutil/replication.go:396]: @nickvanw thanks, I'm curious how you found this?

I'm trying to automate reviews before pushing and this one wasn't noticed by my "team" of reviewers, so perhaps something in my automation needs improving

Comment (timvaillancourt) [reply go/vt/vtctl/reparentutil/replication.go:396]: Update: I tweaked my prompts to review agents to trace control flow, and I now see the same issue 👍

Comment (timvaillancourt) [reply go/vt/vtctl/reparentutil/replication.go:396]: This should be addressed in https://github.com/vitessio/vitess/pull/19069/commits/80eacd8fc4986215f4934ffaa98166cad82cefef

File: go/mysql/replication/replication_status.go [hunk 5]
```diff
@@ -97,6 +97,11 @@ func (s *ReplicationStatus) SQLHealthy() bool {
 	return s.SQLState == ReplicationStateRunning
 }
 
+// IsSemiSyncAcker returns true on tablets that are actively semi-sync replicas.
+func (s *ReplicationStatus) IsSemiSyncAcker() bool {
+	return s.SemiSyncReplicaEnabled && s.SemiSyncReplicaStatus
```
Comment (nickvanw) [inline go/mysql/replication/replication_status.go:108]: `Rpl_semi_sync_replica_status` only tells us whether semi-sync is currently operational on the replica. MySQL turns that status OFF once the receiver thread stops, so after a primary crash a non-GTID replica can have `SemiSyncReplicaEnabled=true` and `SemiSyncReplicaStatus=false` even though it may have been the acker holding the newest committed transaction. In the ERS `After == nil` safety path that means we can silently skip exactly the replica we should fail on. I think this check needs to key off `SemiSyncReplicaEnabled` alone here, or use a separate conservative helper for this fail-safe path. The fake also mirrors status from enabled, so the current tests do not cover this distinction.

**Review by nickvanw — COMMENTED** [review]
I found one more correctness issue in the new non-GTID semi-sync safety path. Details inline.

File: go/mysql/replication/replication_status.go (hunk 5 above)
Comment (timvaillancourt) [reply go/mysql/replication/replication_status.go:108]: @nickvanw wow, this is a great point. That's a real problem - and after digging into this more it doesn't appear MySQL has a realiable way to know if a replica is actively an acker, from the replica standpoint. Apparently the reliable way is from the source/`PRIMARY`, but in reparents that is likely dead 🤔 

At least for this PR, we are checking if a file-based replica is also a semi-sync acker, because this is a case where it wouldn't be safe to ignore a file-based replica. Just checking if the plugin is enabled would be good enough, and I can rename this to `.IsSemiSyncEnabled()`. But, in the future, I was planning on using the `.IsSemiSyncAcker()` function in more ERS code - this discovery means we don't really have an accurate way to determine semi-sync - but again, for this PR it's not too important

What's fascinating is this `func` has been basically unchanged since the beginning of this PR, and no model or human pointed it out until now 🤔

**Comment by timvaillancourt** [comment]
@codex review

**Comment by mhamza15** [comment]
@codex review

File: go/vt/vtctl/reparentutil/util.go [hunk 6]
```diff
@@ -315,16 +315,54 @@ func getValidCandidatesAndPositionsAsList(validCandidates map[string]*RelayLogPo
...
+
 // restrictValidCandidates is used to restrict some candidates from being considered eligible for becoming the intermediate source or the final promotion candidate
-func restrictValidCandidates(validCandidates map[string]*RelayLogPositions, tabletMap map[string]*topo.TabletInfo) (map[string]*RelayLogPositions, error) {
+func restrictValidCandidates(validCandidates map[string]*RelayLogPositions, tabletMap map[string]*topo.TabletInfo, candidateInfoMap map[string]*CandidateInfo,
+	logger logutil.Logger,
+) (map[string]*RelayLogPositions, error) {
+	shardUsesGTID := isGTIDBasedShard(tabletMap, candidateInfoMap)
+	// Fail early if 100% of candidates are not MySQL GTID-based. MySQL GTIDs are a requirement of Vitess here.
+	if len(tabletMap) > 0 && !shardUsesGTID {
+		return nil, vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "shards without MySQL GTID-based candidates are unsupported")
```
[the remark this answers is not in the document]
Comment (timvaillancourt) [reply go/vt/vtctl/reparentutil/util.go:342]: @codex can you explain that in more plain english? How would I reproduce this bug?

[the remark this answers is not in the document]
Comment (timvaillancourt) [reply go/vt/vtctl/reparentutil/util.go:342]: Thanks for the feedback. The early check here is intentional — the normal lifecycle for a blank shard is to call PlannedReparentShard first, which elects the first primary and begins GTID history. ERS is an emergency fallback and should only be called on a shard that already has GTID-based replication established.

If ERS is called with zero/empty GTID positions on every tablet, one of two things is true: either it's being called before PRS has ever run (operator error), or `RESET BINARY LOGS AND GTIDS` was run manually (destructive action requiring human intervention). In both cases, failing fast with `FAILED_PRECONDITION` is the correct behavior — masking the check by deferring it would make these situations harder to diagnose, not easier.

[the remark this answers is not in the document]
Comment (timvaillancourt) [reply go/vt/vtctl/reparentutil/util.go:342]: While inconsequential in reality, I've made a follow-up task to address this hardening in another PR: https://github.com/vitessio/vitess/issues/19902

This PR has become long and has dragged-on, so I don't suggest we complicate this PR with that fix, as it's an edge case that shouldn't be hit

**Comment by timvaillancourt** [comment]
This PR turned out to be more trouble than it's worth

Closing as won't fix. Mixed replication modes will continue to block reparents with an error. The consensus is the user was warned not to do this, so we have no contract to uphold
